include/nbl/video/IGPUDescriptorSetLayout.h

```diff
@@ -25,13 +25,38 @@ class NBL_API IGPUDescriptorSetLayout : public asset::IDescriptorSetLayout<IGPUS
         using base_t = asset::IDescriptorSetLayout<IGPUSampler>;
 
     public:
-        IGPUDescriptorSetLayout(core::smart_refctd_ptr<const ILogicalDevice>&& dev, const SBinding* const _begin, const SBinding* const _end) : base_t(_begin, _end), IBackendObject(std::move(dev)) {}
+        IGPUDescriptorSetLayout(core::smart_refctd_ptr<const ILogicalDevice>&& dev, const SBinding* const _begin, const SBinding* const _end) : base_t(_begin, _end), IBackendObject(std::move(dev))
+        {
+            uint32_t localDescriptorOffset = 0u;
+            for (auto b = _begin; b != _end; ++b)
+            {
+                m_bindingToDescriptorOffsetMap.insert({ b->binding, localDescriptorOffset });
+                localDescriptorOffset += b->count;
+            }
+        }
+
+        inline uint32_t getDescriptorOffsetForBinding(const uint32_t binding) const
+        {
+            auto found = m_bindingToDescriptorOffsetMap.find(binding);
+            if (found != m_bindingToDescriptorOffsetMap.end())
+                return found->second;
+
+            return ~0u;
+        }
 
     protected:
         virtual ~IGPUDescriptorSetLayout() = default;
 
         bool m_isPushDescLayout = false;
         bool m_canUpdateAfterBind = false;
+
+    private:
+        // This maps the descriptor set layout's binding number to the LOCAL offset in the array of descriptors where the descriptors of this binding number start.
+        // Therefore, a given binding number (say, b) having n array elements, will have descriptors stored at the local offset of [m_bindingToDescriptorOffsetMap[b], m_bindingToDescriptorOffsetMap[b]+n).
+        // 
+        // TODO(achal): Can I use a flat array here? It can be done if we enforce any upper limit on VkDescriptorSetLayoutBinding::binding.
+        core::unordered_map<uint32_t, uint32_t> m_bindingToDescriptorOffsetMap;
+
```

Comment:
hoist code from `IEmulatedDescriptorSet`, store a sorted `pair<binding_index_t,offset_t>[EDT_COUNT]`

Reply:
plus one for mutable sampler bindings

Reply:
here's a suggestion on how to build it
```cpp
struct SBindingTempInfo
{
   uint32_t binding;
   uint32_t count;

   inline bool operator<(const SBindingTempInfo& other) const {return binding<other.binding;}
};
struct SBindingRedirect
{
   uint32_t binding;
   uint32_t offset;
};

public:
   IGPUDescriptorSetLayout(core::smart_refctd_ptr<const ILogicalDevice>&& dev, const SBinding* const _begin, const SBinding* const _end) : base_t(_begin, _end), IBackendObject(std::move(dev))
   {
      core::vector<SBindingTempInfo> buildData[EDT_COUNT];
      for (auto b = _begin; b != _end; ++b)
         buildData[b->type].emplace_back(b->binding,b->count);
      std::sort(buildData.begin(),buildData.end());
      // turn counts into offsets by prefix summing
      std::exclusive_sum(buildData.
   }
```

Reply:
again, this should be hoisted to `IDescriptorSetLayout`, so both ICPU and IGPU have this info laying around

Reply:
Resolved.